backends/bmv2/jsonconverter.cpp

```diff
@@ -1216,6 +1216,134 @@ Util::IJson* JsonConverter::convertIf(const CFG::IfNode* node, cstring) {
     return result;
 }
 
+/**
+    Converts a list of P4 defined table entries to JSON.
+
+    Assumes that all the checking on the entries being valid has been
+    done in the validateParsedProgram pass, so this is simply the serialization.
+
+    This function still checks and reports errors in key entry expressions.
+    The JSON generated after a reported error will not be accepted in BMv2.
+
+    @param table A P4 table IR node
+    @param jsonTable the JSON table object to add predefined entries
+
+*/
+void JsonConverter::convertTableEntries(const IR::P4Table *table,
+                                        Util::JsonObject *jsonTable)
+{
+    auto entriesList = table->getEntries();
+    if (entriesList == nullptr) return;
+
+    auto entries = mkArrayField(jsonTable, "entries");
+    int entryPriority = 1; // priority is defined by index position
+    for (auto e : *entriesList->entries) {
+        auto entry = new Util::JsonObject();
+
+        auto keyset = e->getKeys();
+        auto matchKeys = mkArrayField(entry, "match_key");
+        int keyIndex = 0;
+        for (auto k: *keyset->components) {
+            auto key = new Util::JsonObject();
+            auto tableKey = table->getKey()->keyElements->at(keyIndex);
+            unsigned keyWidth = tableKey->expression->type->width_bits();
+            auto matchType = getKeyMatchType(tableKey);
+            key->emplace("match_type", matchType);
+            if (matchType == "valid") {
+                if(k->is<IR::BoolLiteral>())
+                    key->emplace("key", k->to<IR::BoolLiteral>()->toString());
+                else
+                    ::error("%1% invalid 'valid' key expression", k);
+            } else if (matchType == corelib.exactMatch.name) {
+                if(k->is<IR::Constant>())
+                    key->emplace("key", stringRepr(k->to<IR::Constant>()->value, ROUNDUP(keyWidth, 8)));
+                else
+                    ::error("%1% invalid exact key expression", k);
+            } else if (matchType == corelib.ternaryMatch.name) {
+                if (k->is<IR::Mask>()) {
+                    key->emplace("key", stringRepr(k->to<IR::Mask>()->left->to<IR::Constant>()->value, ROUNDUP(keyWidth, 8)));
+                    key->emplace("mask", stringRepr(k->to<IR::Mask>()->right->to<IR::Constant>()->value, ROUNDUP(keyWidth, 8)));
+                } else if (k->is<IR::Constant>()) {
+                    key->emplace("key", stringRepr(k->to<IR::Constant>()->value, ROUNDUP(keyWidth, 8)));
+                    unsigned long mask = (1ULL << keyWidth)-1;
+                    key->emplace("mask", stringRepr(mask, ROUNDUP(keyWidth, 8)));
+                } else if (k->is<IR::DefaultExpression>()) {
```

Comment:
Look at the following example ../testdata/p4_16_samples/select_struct.p4 and check the output from the compiler.